
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#3f51b5">
  <script src="_static/javascripts/modernizr.js"></script>
  
  
  
    <title>Samplesheet &#8212; MpGAP: A generic multi-platform genome assembly pipeline  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/material.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/style.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Manual" href="manual.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" />
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=cyan data-md-color-accent=orange>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#samplesheet" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="index.html" title="MpGAP: A generic multi-platform genome assembly pipeline  documentation"
           class="md-header-nav__button md-logo">
          
            <i class="md-icon">book</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">MpGAP: A generic multi-platform genome assembly pipeline  documentation</span>
          <span class="md-header-nav__topic"> Samplesheet </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder="Search"
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            <a href="https://github.com/fmalmeida/MpGAP" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    MpGAP
  </div>
</a>
          </div>
        </div>
      
      
  
  <script src="_static/javascripts/version_dropdown.js"></script>
  <script>
    var json_loc = ""versions.json"",
        target_loc = "../",
        text = "Versions";
    $( document ).ready( add_version_dropdown(json_loc, target_loc, text));
  </script>
  

    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
          <li class="md-tabs__item"><a href="index.html" class="md-tabs__link">MpGAP: A generic multi-platform genome assembly pipeline  documentation</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="index.html" title="MpGAP: A generic multi-platform genome assembly pipeline documentation" class="md-nav__button md-logo">
      
        <i class="md-icon">book</i>
      
    </a>
    <a href="index.html"
       title="MpGAP: A generic multi-platform genome assembly pipeline documentation">MpGAP: A generic multi-platform genome assembly pipeline  documentation</a>
  </label>
    <div class="md-nav__source">
      <a href="https://github.com/fmalmeida/MpGAP" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    MpGAP
  </div>
</a>
    </div>
  
  

  
  <ul class="md-nav__list">
    <li class="md-nav__item">
    
    
      <a href="installation.html" class="md-nav__link">Installation</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="quickstart.html" class="md-nav__link">Quickstart</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    <label class="md-nav__link md-nav__link--active" for="__toc"> Samplesheet </label>
    
      <a href="#" class="md-nav__link md-nav__link--active">Samplesheet</a>
      
        
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#samplesheet--page-root" class="md-nav__link">Samplesheet</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#samplesheet-header" class="md-nav__link">Samplesheet header</a>
        </li>
        <li class="md-nav__item"><a href="#sample-identification" class="md-nav__link">Sample identification</a>
        </li>
        <li class="md-nav__item"><a href="#yaml-keys-related-to-input-files" class="md-nav__link">YAML keys related to input files</a>
        </li>
        <li class="md-nav__item"><a href="#yaml-keys-related-to-configuration" class="md-nav__link">YAML keys related to configuration</a>
        </li>
        <li class="md-nav__item"><a href="#complete-samplesheet-example" class="md-nav__link">Complete samplesheet example</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="_sources/samplesheet.rst.txt">Show Source</a> </li>

  </ul>
</nav>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="manual.html" class="md-nav__link">Manual</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="config.html" class="md-nav__link">Configuration File</a>
      
    
    </li>
  </ul>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#samplesheet--page-root" class="md-nav__link">Samplesheet</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#samplesheet-header" class="md-nav__link">Samplesheet header</a>
        </li>
        <li class="md-nav__item"><a href="#sample-identification" class="md-nav__link">Sample identification</a>
        </li>
        <li class="md-nav__item"><a href="#yaml-keys-related-to-input-files" class="md-nav__link">YAML keys related to input files</a>
        </li>
        <li class="md-nav__item"><a href="#yaml-keys-related-to-configuration" class="md-nav__link">YAML keys related to configuration</a>
        </li>
        <li class="md-nav__item"><a href="#complete-samplesheet-example" class="md-nav__link">Complete samplesheet example</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="_sources/samplesheet.rst.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  <section id="samplesheet">
<span id="id1"></span><h1 id="samplesheet--page-root">Samplesheet<a class="headerlink" href="#samplesheet--page-root" title="Permalink to this headline">¶</a></h1>
<p>The samplesheet is a required YAML document that is used to describe the input samples and, if desired, its “sample-specific” configuration. The input samplesheet is given using the <code class="docutils literal notranslate"><span class="pre">--input</span></code> parameter.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>A samplesheet template can be downloaded with: <code class="docutils literal notranslate"><span class="pre">nextflow</span> <span class="pre">run</span> <span class="pre">fmalmeida/mpgap</span> <span class="pre">--get_samplesheet</span></code></p>
</div>
<p>A guide on how to proper configure it is shown below:</p>
<section id="samplesheet-header">
<h2 id="samplesheet-header">Samplesheet header<a class="headerlink" href="#samplesheet-header" title="Permalink to this headline">¶</a></h2>
<p>The first line of the file must be the header followed by an indentation (two white spaces):</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">samplesheet</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">...</span><span class="p">:</span><span class="w"></span>
</pre></div>
</div>
</section>
<section id="sample-identification">
<h2 id="sample-identification">Sample identification<a class="headerlink" href="#sample-identification" title="Permalink to this headline">¶</a></h2>
<p>Each sample must be identified by the tag <code class="docutils literal notranslate"><span class="pre">id</span></code> in the YAML file, followed by the sample’s input tags (YAML keys) that shall be used by the pipeline:</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This value will be used to create sub-directories in the output directory. Thus, to not use white spaces.</p>
</div>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">samplesheet</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_1</span><span class="w"></span>
<span class="w">    </span><span class="nt">...</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">...</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_2</span><span class="w"></span>
<span class="w">    </span><span class="nt">...</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">...</span><span class="p">:</span><span class="w"></span>
</pre></div>
</div>
</section>
<section id="yaml-keys-related-to-input-files">
<h2 id="yaml-keys-related-to-input-files">YAML keys related to input files<a class="headerlink" href="#yaml-keys-related-to-input-files" title="Permalink to this headline">¶</a></h2>
<p>These are the tags that are used to represent/set the input files that shall be used for each sample. The available tags are:</p>
<table>
<colgroup>
<col style="width: 25%"/>
<col style="width: 75%"/>
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Input tags</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">illumina</span></code></p></td>
<td><p>Used to set path to illumina raw reads (paired, unpaired or both).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pacbio</span></code></p></td>
<td><p>Used to set path to pacbio raw reads (mutually excludable with <code class="docutils literal notranslate"><span class="pre">nanopore</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pacbio_bam</span></code></p></td>
<td><p>Used to set path to pacbio bam file (used in conjunction with <code class="docutils literal notranslate"><span class="pre">pacbio</span></code> for long reads assembly polishing with gcpp)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">nanopore</span></code></p></td>
<td><p>Used to set path to nanopore raw reads (mutually excludable with <code class="docutils literal notranslate"><span class="pre">pacbio</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nanopolish_fast5</span></code></p></td>
<td><p>Used to set path to nanopore raw FAST5 data (used in conjunction with <code class="docutils literal notranslate"><span class="pre">nanopore</span></code> for long reads assembly polishing with Nanopolish)</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note for the illumina tag/key.</p>
<ul class="simple">
<li><p>When using both paired and unpaired reads, the paired reads must be given first, in the order: pair 1, pair 2, unpaired.</p></li>
<li><p>Otherwise, if using only paired reads, they must be given in the order: pair 1, pair 2.</p></li>
<li><p>If using only unpaired reads, only one entry is expected. Check samples in the template to 1, 4 and 5 to understand it.</p></li>
<li><dl class="simple">
<dt>The illumina tag is the only one that <strong>must</strong> be set in indented newlines</dt><dd><ul>
<li><p>two white spaces relative to the</p></li>
<li><p>one line per read as shown in the complete samplesheet example.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>All the other input tags <strong>must</strong> be set in the same line, right after the separator (“:”), without quotations, white spaces or signs.</p>
</div>
</section>
<section id="yaml-keys-related-to-configuration">
<h2 id="yaml-keys-related-to-configuration">YAML keys related to configuration<a class="headerlink" href="#yaml-keys-related-to-configuration" title="Permalink to this headline">¶</a></h2>
<p>These are the tags that are used to represent/set the “sample-specific” assembly configuration that shall be used for each sample.</p>
<p>By default, if not set inside the samplesheet, the pipeline will use the configurations set via the “nextflow config file” or via the command line. Otherwise, if set inside the samplesheet, they will overwrite the pipeline’s configuration for that specific sample.</p>
<p>Please, the <a class="reference internal" href="manual.html#manual"><span class="std std-ref">manual reference page</span></a> the global/defaults configurations.</p>
<p>The available tags are:</p>
<table>
<colgroup>
<col style="width: 40%"/>
<col style="width: 60%"/>
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Input tags</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">hybrid_strategy</span></code></p></td>
<td><p>This sets which strategy to run when performing hybrid assemblies. Please read the <a class="reference internal" href="manual.html#manual"><span class="std std-ref">Manual</span></a> reference page to understand the adopted strategies. Options are: <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, or <code class="docutils literal notranslate"><span class="pre">both</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">corrected_long_reads</span></code></p></td>
<td><p>Tells whether the long reads used are corrected or not. Options: <code class="docutils literal notranslate"><span class="pre">true</span></code>, <code class="docutils literal notranslate"><span class="pre">false</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nanopolish_max_haplotypes</span></code></p></td>
<td><p>It sets the max number of haplotypes to be considered by Nanopolish. Sometimes the pipeline may crash because to much variation was found exceeding the limit. Options: any integer value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">medaka_model</span></code></p></td>
<td><p>Used to polish a longreads-only assembly with Medaka. It selects a Medaka ONT sequencing model for polishing. Please read <a class="reference external" href="https://github.com/nanoporetech/medaka#models">medaka manual</a> to know the available models.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">shasta_config</span></code></p></td>
<td><p>This selects the shasta configuration file to be used when assembling reads. It is now mandatory for shasta since its v0.8 release. Please read the <a class="reference external" href="https://chanzuckerberg.github.io/shasta/Configurations.html">shasta configuration manual page</a> to know the available models.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">genome_size</span></code></p></td>
<td><p>This sets the expected genome sizes for canu, wtdbg2 and haslr assemblers, which require this value. Options are estimatives with common suffices, for example: <code class="docutils literal notranslate"><span class="pre">3.7m</span></code>, <code class="docutils literal notranslate"><span class="pre">2.8g</span></code>, etc.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">wtdbg2_technology</span></code></p></td>
<td><p>This sets the technology of input reads. It is required by wtdbg2. Options are: <code class="docutils literal notranslate"><span class="pre">ont</span></code> for Nanopore reads, <code class="docutils literal notranslate"><span class="pre">rs</span></code> for PacBio RSII, <code class="docutils literal notranslate"><span class="pre">sq</span></code> for PacBio Sequel, <code class="docutils literal notranslate"><span class="pre">ccs</span></code> for PacBio CCS reads.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="complete-samplesheet-example">
<h2 id="complete-samplesheet-example">Complete samplesheet example<a class="headerlink" href="#complete-samplesheet-example" title="Permalink to this headline">¶</a></h2>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># This is an exemplification of the accepted YAML syntax accepted by the MpGAP pipeline</span><span class="w"></span>
<span class="c1">#</span><span class="w"></span>
<span class="c1"># The file must contain a header-line with the key "samplesheet:". All the</span><span class="w"></span>
<span class="c1"># input samples must be given nested to this key. Obs: The header-line and</span><span class="w"></span>
<span class="c1"># the nest indentation (two white spaces) are REQUIRED.</span><span class="w"></span>
<span class="c1">#</span><span class="w"></span>
<span class="c1"># Each sample must initiate with the key "id:" listed with a "-". This value</span><span class="w"></span>
<span class="c1"># is what will be used as prefix for outputs. The input keys must be given </span><span class="w"></span>
<span class="c1"># right after this id key, with the same indent without the "-".</span><span class="w"></span>
<span class="c1">#</span><span class="w"></span>
<span class="c1"># Please read the https://mpgap.readthedocs.io/en/latest/samplesheet.html documentation page</span><span class="w"></span>
<span class="c1"># in order to understand how to properly set it up and which are the available/expected keys</span><span class="w"></span>
<span class="c1">#</span><span class="w"></span>
<span class="c1"># A template (with the correct fields, syntax and indentation) is given below:</span><span class="w"></span>

<span class="nt">samplesheet</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_1</span><span class="w"></span>
<span class="w">    </span><span class="nt">illumina</span><span class="p">:</span><span class="w"> </span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_1.fastq.gz</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_2.fastq.gz</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_2</span><span class="w"></span>
<span class="w">    </span><span class="nt">illumina</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_unpaired.fastq.gz</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_3</span><span class="w"></span>
<span class="w">    </span><span class="nt">illumina</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_1.fastq.gz</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_2.fastq.gz</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/pairs_merged.fastq.gz</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_4</span><span class="w"></span>
<span class="w">    </span><span class="nt">nanopore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/ont_reads.fastq.gz</span><span class="w"></span>
<span class="w">    </span><span class="nt">corrected_long_reads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"></span>
<span class="w">    </span><span class="nt">nanopolish_fast5</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/kleb/fast5_pass</span><span class="w"></span>
<span class="w">    </span><span class="nt">genome_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5.5m</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_5</span><span class="w"></span>
<span class="w">    </span><span class="nt">pacbio</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/pacbio_reads.fastq.gz</span><span class="w"></span>
<span class="w">    </span><span class="nt">pacbio_bam</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/pacbio_reads.subreads.bam</span><span class="w"></span>
<span class="w">    </span><span class="nt">wtdbg2_technology</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">rs</span><span class="w">  </span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sample_6</span><span class="w"></span>
<span class="w">    </span><span class="nt">illumina</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_1.fastq.gz</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/reads_2.fastq.gz</span><span class="w"></span>
<span class="w">    </span><span class="nt">nanopore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset/ont_reads.fastq.gz</span><span class="w"></span>
<span class="w">    </span><span class="nt">hybrid_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">both</span><span class="w"></span>
</pre></div>
</div>
</section>
</section>


          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
            <a href="quickstart.html" title="Quickstart"
               class="md-flex md-footer-nav__link md-footer-nav__link--prev"
               rel="prev">
              <div class="md-flex__cell md-flex__cell--shrink">
                <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
              </div>
              <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
                <span class="md-flex__ellipsis">
                  <span
                      class="md-footer-nav__direction"> Previous </span> Quickstart </span>
              </div>
            </a>
          
          
            <a href="manual.html" title="Manual"
               class="md-flex md-footer-nav__link md-footer-nav__link--next"
               rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span
                class="md-flex__ellipsis"> <span
                class="md-footer-nav__direction"> Next </span> Manual </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink"><i
                class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2019, fmalmeida/mpgap: A genome assembly pipeline for multiple sequencing platforms..
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 4.2.0.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>